pkg/ssh/mux/muxer.go

```diff
@@ -0,0 +1,147 @@
+package mux
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/gliderlabs/ssh"
+	"github.com/pkg/errors"
+)
+
+var (
+	// ContextCmd points to the command-line in the Context
+	ContextCmd = contextType("context")
+	// ContextPattern points to the pattern that matched the Handler
+	ContextPattern = contextType("pattern")
+)
+
+type (
+	contextType string
+	// Muxer is a simple SSH Session muxer whose Handlers
+	//  are matched against a regexp-pattern.
+	//  Middlewares can be injected as well.
+	Muxer interface {
+		// Use, inserts one or more middleware(s).
+		//  They are executed in FIFE (First In, First Executed).
+		//  This means the last one in the list is the one executing
+		//  closest to the Handler.
+		Use(serv ...MiddlewareFunc)
+
+		// AddHandler to the Muxer
+		//  This will wrap the Handler in a HandlerFunc
+		AddHandler(pattern string, serv Handler) error
+
+		// AddHandlerFunc to the Muxer
+		AddHandlerFunc(pattern string, serv HandlerFunc) error
+
+		// Handle a given session. This should be given to gliderlabs/ssh.Server.Handler()
+		Handle() ssh.Handler
+	}
+
+	muxer struct {
+		handlers map[string]HandlerFunc
+		regexps  map[string]*regexp.Regexp
+		mws      []MiddlewareFunc
+	}
+)
+
+// New creates a new Muxer
+func New() Muxer {
+	return &muxer{
+		handlers: make(map[string]HandlerFunc),
+		regexps:  make(map[string]*regexp.Regexp),
+	}
+}
+
+func (s *muxer) Use(serv ...MiddlewareFunc) {
+	// We prepend the middlewares instead, so that
+	//  Use(Foo, Bar) results in Foo(Bar(Handler))
+	//  being ran, instead of Bar(Foo(Handler))
+	// NOTE: This is probably somewhat inefficient
+	//  (but only ran at setup, so who cares)
+	for _, mw := range serv {
+		s.mws = append([]MiddlewareFunc{mw}, s.mws...)
+	}
+}
+
+func (s *muxer) AddHandler(pattern string, serv Handler) error {
+	return s.AddHandlerFunc(pattern, func(c context.Context, s ssh.Session) error {
+		return serv.Execute(c, s)
+	})
```

Comment:
Looking at how `http.HandlerFunc` works, seems like this could be part of the type :thinking: